### DISCUSSION TITLE
How to avoid VMs to go to sleep automatically?

### DISCUSSION THREAD

**Discussion by yajo** [vagrant-libvirt/vagrant-libvirt#1348] [Q&A]
I managed to boot machines with vagrant-libvirt. I'm using them with molecule-libvirt to test ansible playbooks.

However, at some point of the playbook (it's a bit long), VMs go to sleep automatically. The workaround I've found so far is to open gnome-boxes and keep them open, so they won't go to sleep, but that's not much comfortable.

Is there a way to make VMs stay awake? At least for a longer timeout?

**Comment by electrofelix** [comment]
At a guess I'd suspect that something (likely systemd) is checking for active consoles in the VM and suspending automatically after some timeout. I've seen this locally with a machine when I use a mouse keyboard switcher between them, the one without the keyboard/mouse input will eventually hibernate.

Can't be completely sure as I've only just made the same change locally, but adding the following file in the guest might prevent the shutdown:
```
sudo mkdir -p /etc/systemd/sleep.conf.d/
cat <<EOF | sudo tee /etc/systemd/sleep.conf.d/no-suspend-hibernate.conf >/dev/null
[Sleep]
# disable hibernation
# doc : https://archived.forum.manjaro.org/t/turn-off-disable-hibernate-completely/139939
AllowSuspend=no
AllowHibernation=no
AllowHybridSleep=no
AllowSuspendThenHibernate=no
EOF
```

I would have assumed the default settings for the VM should prevent it though, as it should have a console connected along with a keyboard and mouse, which should simulate an active terminal. But maybe that is insufficient these days with what settings are present in systemd by default as there would be no input coming in from either.

**Reply by yajo** [reply]
Sadly I have tried your suggestion and VMs still go to sleep after a few minutes.

**Comment by yajo** [comment]
The workaround I found is to open gnome-boxes and tick the box to allow them to run in background:

![pantallazo_2021-09-17_12-41](https://user-images.githubusercontent.com/973709/133772365-3836238a-8239-4be2-bc44-1065116010b4.png)

I don't know what does that do in the background, or if that setting can be enabled through vagrant-libvirt automatically, or a script... If anybody knows, it'd be great to know!

**Reply by electrofelix** [reply]
Are you using flatpak? Wondering if you are bumping into https://gitlab.gnome.org/GNOME/gnome-boxes/-/issues/379 where the libvirt instance in your session is inside another container process that is suspending the running VMs?

**Reply by yajo** [reply]
No, I installed it in the host.

**Comment by electrofelix** [comment]
So I'm still stumped on what the cause is here, obviously for session based machines, some work has been done on automatically suspending any running unless they are marked to allow them to remain running. I suspect the quickest way will be to talk to the gnome-boxes developers.

**Comment by trinitronx** [comment]
For Linux instances with SystemD (if sleep is being triggered by the init system's targets)
```
## To list all SystemD targets (not just sleep-related)
sudo systemctl list-units --type target

## Mask Sleep-related targets (disables by symlinking to /dev/null)
sudo systemctl mask sleep.target suspend.target hibernate.target hybrid-sleep.target
```

For Linux systems where the Gnome desktop environment is causing sleep:

```
# View ALL power management plugin settings in Gnome settings:
gsettings list-recursively org.gnome.settings-daemon.plugins.power
## View ALL gnome.settings-daemon.plugins matching 'sleep'
gsettings list-recursively org.gnome.settings-daemon.plugins  | grep -i sleep 

## Turn off sleep
gsettings set org.gnome.settings-daemon.plugins.power sleep-inactive-ac-timeout '0'
gsettings set org.gnome.settings-daemon.plugins.power sleep-inactive-battery-timeout '0'
```

A VM going to sleep is also a problem with macOS instances too.
Although it's not related to the OP's Linux-based VM, here's the appropriate macOS command to disable sleep:

```
## Check Sleep Settings
pmset -g

## Disable all the sleep settings
for setting in sleep hibernatemode displaysleep powernap disksleep ; do
  sudo pmset -a $setting 0
done

sudo pmset -a disablesleep 1
```
